### PR TITLE
[API]: Add drain-aware RoleSet rollout and scale-in

### LINKED ISSUE #2413: [Feature] Support drain-aware RoleSet rollout and scale-in
## Background

Currently, RoleSet deletes pods directly during rollout or scale-in.

For LLM serving workloads, this can interrupt in-flight requests, especially when:

- streaming decode requests are still producing tokens;
- decode pods in P/D disaggregation still have active streams;
- SGLang async prefill still has unfinished prefill requests;
- the gateway may continue routing new requests to a pod that is about to be deleted.

This can cause long-tail request failures, broken streams, and inconsistent prefill/decode state.

## Goal

RoleSet should support a simple time-based drain window before deleting pods:

1. The RoleSet controller selects a pod to delete.
2. Instead of deleting it immediately, the controller marks the pod as draining.
3. The gateway/router observes the draining state and stops routing new requests to the pod.
4. The controller waits for the configured `drain.timeoutSeconds`.
5. After the wait window expires, the controller deletes the pod.

This should work for both direct `RoleSet` users and the more common `StormService` users.

## API Design

Add an optional `Drain` field to `RoleSpec`.

Because `StormService` reuses `RoleSetSpec` through `spec.template.spec`, this field can be configured in both:

- `RoleSet.spec.roles[].drain`
- `StormService.spec.template.spec.roles[].drain`

### Go struct

```go
type RoleSpec struct {
    Name string `json:"name,omitempty"`

    // Replicas is the number of desired replicas.
    // +optional
    Replicas *int32 `json:"replicas,omitempty"`

    // Drain configures a time-based drain window before deleting pods
    // during rollout or scale-in.
    //
    // When unset, or when TimeoutSeconds is nil/0, RoleSet keeps the
    // current behavior and deletes pods immediately.
    //
    // When TimeoutSeconds > 0, RoleSet marks the selected pod as draining,
    // waits for the configured timeout, and then deletes the pod.
    //
    // +optional
    Drain *RoleDrainSpec `json:"drain,omitempty"`

    // ... existing fields ...
}
```

```go
// RoleDrainSpec configures drain-aware deletion for pods belonging to a role.
type RoleDrainSpec struct {
    // TimeoutSeconds is the time-based drain window before deleting a pod.
    //
    // If unset or 0, drain-aware deletion is disabled.
    // If greater than 0, the RoleSet controller marks the pod as draining,
    // waits this duration, and then deletes the pod.
    //
    // +optional
    // +kubebuilder:validation:Minimum=0
    TimeoutSeconds *int32 `json:"timeoutSeconds,omitempty"`
}
```

## StormService Example

```yaml
apiVersion: orchestration.aibrix.ai/v1alpha1
kind: StormService
metadata:
  name: llama-pd
spec:
  replicas: 1
  selector:
    matchLabels:
      app: llama-pd
  template:
    metadata:
      labels:
        app: llama-pd
    spec:
      roles:
      - name: prefill
        replicas: 2
        drain:
          timeoutSeconds: 30
        template:
          spec:
            containers:
            - name: engine
              image: example/prefill:latest
      - name: decode
        replicas: 4
        drain:
          timeoutSeconds: 120
        template:
          spec:
            containers:
            - name: engine
              image: example/decode:latest
```

## RoleSet Example

```yaml
apiVersion: orchestration.aibrix.ai/v1alpha1
kind: RoleSet
metadata:
  name: llama-pd-rs
spec:
  roles:
  - name: decode
    replicas: 4
    drain:
      timeoutSeconds: 120
    template:
      spec:
        containers:
        - name: engine
          image: example/decode:latest
```

## Field Semantics

- If `drain` is unset, existing behavior is preserved and pods are deleted without waiting.
- If `drain.timeoutSeconds` is unset or `0`, existing behavior is preserved.
- If `drain.timeoutSeconds > 0`, drain-aware deletion is enabled.
- Before deleting a pod, the controller marks it as draining and waits for the configured timeout.
- After the timeout expires, the controller deletes the pod to avoid blocking rollout or scale-in forever.

## Draining Annotations

When a pod is selected for deletion, the RoleSet controller patches the pod with:

```yaml
aibrix.ai/draining: "true"
aibrix.ai/drain-start-time: "<RFC3339 timestamp>"
aibrix.ai/drain-reason: "scale-in" # or "rollout"
aibrix.ai/drain-target-action: "delete" # future values may include "sleep"
```

These annotations are used to:

- let the gateway/router identify draining pods;
- let the controller recover drain state after restart;
- help users debug why a pod has not been deleted yet;
- let future lifecycle controllers distinguish the post-drain target action without adding action fields to the `drain` API.

## Drain Target Action

The first version only uses drain-aware deletion, so `aibrix.ai/drain-target-action` should be set to `delete`.

The annotation intentionally keeps the CRD API small: `drain` describes the pre-action drain window, while the controller-owned annotation describes what happens after draining. This leaves room for a future sleep/warm-standby controller to reuse the same draining protocol by setting:

```yaml
aibrix.ai/drain-reason: "sleep"
aibrix.ai/drain-target-action: "sleep"
```

The gateway should treat all `aibrix.ai/draining=true` pods as unavailable for new requests, regardless of the target action.

## Gateway Behavior

The gateway/router should filter out pods with:

```yaml
aibrix.ai/draining: "true"
```

Pods with this annotation should not receive new requests.

The first version does not require the gateway to report active request counts or write back drain completion state.

## Implementation Scope

Drain-aware deletion must cover both deletion layers:

1. RoleSet controller / role syncers
   - For `podGroupSize <= 1`, RoleSet deletes Pods directly during scale-in and rollout.
   - These direct Pod deletions must patch drain annotations and wait for `drain.timeoutSeconds` before deleting the Pod.

2. PodSet path
   - For `podGroupSize > 1`, RoleSet deletes PodSet objects during scale-in and rollout.
   - PodSet controller owns the actual Pods and deletes them during finalization, recreate, unhealthy replacement, and scale-down.
   - The drain config from `RoleSpec` must be propagated to the generated PodSet, and PodSet controller must apply the same drain-before-delete behavior before deleting member Pods.

The user-facing API should remain `RoleSpec.drain`; PodSet can carry this as controller-owned internal state, such as a spec field or annotation, so direct users do not need to configure PodSet drain separately.

## Controller Behavior

When the RoleSet controller is about to delete a pod:

1. If the role does not configure `drain.timeoutSeconds`, delete the pod using the existing behavior.
2. If the pod does not have `aibrix.ai/draining=true`:
   - patch the draining annotations with `aibrix.ai/drain-target-action=delete`;
   - set `aibrix.ai/drain-start-time`;
   - do not delete the pod in the current reconcile;
   - requeue after the drain timeout.
3. If the pod is already draining:
   - read `aibrix.ai/drain-start-time`;
   - if the timeout has not expired, keep waiting;
   - if the timeout has expired, delete the pod.
4. If the annotation is missing or malformed, the controller should record an event and conservatively restart the drain flow or keep waiting.

## Non-goals

The first version does not implement:

- controller calls to sidecar or engine drain endpoints;
- controller calls to vLLM sleep/wake endpoints;
- controller polling of active streams or active prefill requests;
- gateway write-back of drain completion state;
- shared active request counters across gateway replicas;
- a `PodDrain` CRD;
- early deletion based on actual request completion.

The first version only provides a time-based drain window to reduce the chance of interrupting long-running requests during rollout or scale-in.

## Rollout / Scale-in Impact

Enabling this feature increases rollout and scale-in duration.

The additional duration is roughly:

```text
extra duration ~= number of drain batches * drain.timeoutSeconds
```

For example, if a decode role has 10 pods, `maxUnavailable=1`, and `drain.timeoutSeconds=60`:

```text
extra duration ~= 10 * 60s = 600s
```

Therefore, this feature should be disabled by default and enabled explicitly per role.

## Acceptance Criteria

- If `drain.timeoutSeconds` is not configured, RoleSet behavior remains unchanged.
- StormService users can enable this through `spec.template.spec.roles[].drain.timeoutSeconds`.
- RoleSet users can enable this through `spec.roles[].drain.timeoutSeconds`.
- During scale-in, pods selected for removal are marked as draining before deletion.
- During rollout, old pods are marked as draining before deletion.
- The gateway does not route new requests to draining pods.
- The controller can recover drain state from pod annotations after restart.
- Pods are deleted after the drain timeout expires.
- The controller emits clear events/logs for drain state transitions.

### PR DESCRIPTION
**Pull request by googs1025**
## Summary
- add optional RoleSpec.drain and internal PodSetSpec.drain API support
- mark pods as draining before controller-managed deletion and delete after timeout
- filter draining pods from gateway ready routing and wire drain-aware deletion through RoleSet/PodSet lifecycle paths
- regenerate deepcopy, CRDs, chart CRDs, and applyconfiguration helpers
- add webhook/admission, controller integration, e2e, and unit coverage for drain behavior

Fixes #2413

## Controller Flow
```mermaid
flowchart TD
    A[RoleSet reconcile] --> B[Build desired role membership]
    B --> C{Need to delete pod?}
    C -- No --> Z[Continue normal sync]
    C -- Yes --> D{role.drain.timeoutSeconds > 0?}
    D -- No --> E[Delete pod immediately]
    D -- Yes --> F{Pod already has aibrix.ai/draining=true?}
    F -- No --> G[Patch pod drain annotations]
    G --> H[Record PodDrainStarted event]
    H --> I[Requeue after drain timeout]
    F -- Yes --> J{Drain state valid?}
    J -- No --> K[Reset drain annotations]
    K --> L[Record PodDrainStateInvalid warning]
    L --> I
    J -- Yes --> M{Timeout elapsed?}
    M -- No --> N[V-log waiting state and requeue]
    M -- Yes --> O[Delete pod]
    O --> P[Record PodDrainCompleted event]
```

## Gateway Interaction
```mermaid
sequenceDiagram
    participant RS as RoleSet/PodSet controller
    participant Pod as Serving Pod
    participant Cache as Gateway pod cache
    participant GW as Gateway router

    RS->>Pod: Patch aibrix.ai/draining=true
    RS->>Pod: Patch drain-start-time, drain-reason, drain-target-action=delete
    Pod-->>Cache: Pod update observed
    GW->>Cache: List ready pods for routing
    Cache-->>GW: Return ready pods excluding draining=true
    GW-->>GW: Route only to non-draining ready pods
    RS-->>RS: Wait until drain timeout expires
    RS->>Pod: Delete pod
```

## Drain Annotations
- `aibrix.ai/draining=true`: pod is excluded from gateway ready routing and is in controller-managed drain.
- `aibrix.ai/drain-start-time`: RFC3339 UTC start time used by the controller timeout check.
- `aibrix.ai/drain-reason`: diagnostic reason, currently `scale-in` or `rollout`.
- `aibrix.ai/drain-target-action=delete`: guards the delete path from acting on stale or foreign drain state.

## Tests
- env GOTOOLCHAIN=go1.23.12 GOCACHE=/private/tmp/aibrix-gocache go test ./test/e2e -run `^TestRoleSetDrainScaleDown$`
- env GOTOOLCHAIN=go1.23.12 GOCACHE=/private/tmp/aibrix-gocache go test ./test/integration/webhook ./test/integration/controller -run `^$`
- env GOTOOLCHAIN=go1.23.12 GOCACHE=/private/tmp/aibrix-gocache go test ./pkg/controller/drain ./pkg/controller/roleset ./pkg/controller/podset ./pkg/utils -run `Test(DeletePods|StatelessRoleSyncer|StatefulRoleSyncer|PodSetRoleSyncer|HandleScaleDown|CreatePodFromTemplate|FilterReadyPod)`
- env GOTOOLCHAIN=go1.23.12 GOCACHE=/Users/zhenyu.jiang/go/src/golanglearning/new_project/aibrix/.gocache bash ./hack/verify-codegen.sh
- hack/verify-crd-sync.sh
- env GOCACHE=/Users/zhenyu.jiang/go/src/golanglearning/new_project/aibrix/.gocache go test ./pkg/controller/drain ./pkg/controller/podset ./pkg/utils ./pkg/client/applyconfiguration/...
- git diff --check remote/main...HEAD

## Notes
- Full focused RoleSet/webhook envtest execution was attempted but blocked locally while starting the envtest control plane; compile-level integration checks pass with `-run ^$`.
- E2E drain coverage is gated by `AIBRIX_ROLESET_DRAIN_E2E=true`, following existing RoleSet e2e patterns.
- Comet artifacts and handoff.md are intentionally excluded from this PR.

### CODE REVIEW

File: pkg/controller/roleset/rolling.go [hunk 1]
```diff
@@ -32,28 +33,28 @@ import (
...
 		}
-		scaling = scaling || s
+		result.Merge(scaleResult)
 	}
-	if scaling {
+	if result.Changed {
 		klog.Infof("[RollingManagerSequential.Next] waiting for roleset %s/%s to be scaled", roleSet.Namespace, roleSet.Name)
-		return nil
+		return result, nil
 	}
```
[the remark this answers is not in the document]
Comment (googs1025) [reply pkg/controller/roleset/rolling.go]: Fixed. Sequential, Parallel, and Interleave rolling managers now wait when either result.Changed is true or result.RequeueAfter is greater than zero, so drain-in-progress scale-down does not proceed into rollout early.

File: pkg/controller/podset/podset_controller.go [hunk 2]
```diff
@@ -285,69 +288,72 @@ func (r *PodSetReconciler) handleReplaceUnhealthy(ctx context.Context, podSet *o
...
+	return controllerdrain.Result{}, nil
 }
 
 func (r *PodSetReconciler) handleScaleDown(ctx context.Context, podSet *orchestrationv1alpha1.PodSet,
-	activePods []corev1.Pod, currentPodCount, desiredPodCount int) error {
+	activePods []corev1.Pod, currentPodCount, desiredPodCount int) (controllerdrain.Result, error) {
 
 	podsToDelete := currentPodCount - desiredPodCount
 	r.EventRecorder.Eventf(podSet, corev1.EventTypeNormal, "ScalingDown",
 		"Deleting %d excess pods to meet desired count of %d", podsToDelete, desiredPodCount)
```
[the remark this answers is not in the document]
Comment (googs1025) [reply pkg/controller/podset/podset_controller.go]: Fixed. PodSet scale-down now emits a drain-specific message when drain.timeoutSeconds is active and keeps the delete wording for immediate deletion.

File: pkg/controller/podset/podset_controller.go [hunk 3]
```diff
@@ -235,34 +241,31 @@ func (r *PodSetReconciler) reconcilePods(ctx context.Context, podSet *orchestrat
...
 		}
 	}
 	if len(podsToDelete) > 0 {
 		klog.InfoS("Found unhealthy pods to be deleted", "count", len(podsToDelete), "podset", podSet.Name)
-		for _, pod := range podsToDelete {
-			if err := r.Delete(ctx, pod); err != nil {
-				klog.ErrorS(err, "Failed to delete unhealthy pod", "pod", pod.Name)
-				continue
-			}
+		result, err := controllerdrain.DeletePods(ctx, r.Client, r.EventRecorder, podSet, podsToDelete, podSet.Spec.Drain, aibrixconst.PodDrainReasonRollout, time.Now())
```
[the remark this answers is not in the document]
Comment (googs1025) [reply pkg/controller/podset/podset_controller.go]: Fixed. PodSet ReplaceUnhealthy now passes nil drain config to DeletePods so unhealthy Pods are deleted immediately and replacement is not blocked by drain timeout. Added TestHandleReplaceUnhealthyDeletesImmediatelyWithoutDrain.

File: pkg/controller/drain/drain.go [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+		annotations = map[string]string{}
+	} else {
+		annotations = copyStringMap(annotations)
+	}
+	annotations[aibrixconst.PodDrainingAnnotationKey] = "true"
+	annotations[aibrixconst.PodDrainStartTimeAnnotationKey] = now.UTC().Format(time.RFC3339)
+	annotations[aibrixconst.PodDrainReasonAnnotationKey] = reason
+	annotations[aibrixconst.PodDrainTargetActionAnnotationKey] = aibrixconst.PodDrainTargetActionDelete
+	pod.SetAnnotations(annotations)
+	if err := cli.Patch(ctx, pod, client.MergeFrom(before)); err != nil {
```
[the remark this answers is not in the document]
Comment (googs1025) [reply pkg/controller/drain/drain.go:202]: Fixed. startDrain now treats Patch NotFound as a concurrent deletion no-op, matching the delete path behavior. Added TestDeletePodsIgnoresPatchNotFound.

File: pkg/controller/drain/drain.go [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+	orchestrationv1alpha1 "github.com/vllm-project/aibrix/api/orchestration/v1alpha1"
+	aibrixconst "github.com/vllm-project/aibrix/pkg/constants"
+)
+
+const (
+	EventStarted      = "PodDrainStarted"
+	EventCompleted    = "PodDrainCompleted"
+	EventStateInvalid = "PodDrainStateInvalid"
+
+	maxFutureStartSkew = 30 * time.Second
```
[the remark this answers is not in the document]
Comment (googs1025) [reply pkg/controller/drain/drain.go]: Fixed. Future drain start times are now treated as elapsed=0 instead of invalid state, so the controller waits without resetting annotations or emitting PodDrainStateInvalid repeatedly. Added TestDeletePodsTreatsFutureStartTimeAsWaitingWithoutEvent.

File: pkg/controller/drain/drain.go [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+	if rawStart == "" {
+		return time.Time{}, false, "missing drain start time"
+	}
+	startTime, err := time.Parse(time.RFC3339, rawStart)
+	if err != nil {
+		return time.Time{}, false, fmt.Sprintf("malformed drain start time %q", rawStart)
+	}
+	if startTime.After(now.Add(maxFutureStartSkew)) {
+		return time.Time{}, false, "drain start time is too far in the future"
+	}
```
[the remark this answers is not in the document]
Comment (googs1025) [reply pkg/controller/drain/drain.go]: Fixed. Future drain start times are now treated as elapsed=0 instead of invalid state, so the controller waits without resetting annotations or emitting PodDrainStateInvalid repeatedly. Added TestDeletePodsTreatsFutureStartTimeAsWaitingWithoutEvent.

**Comment by varungup90** [comment]
### **Feedback & Code Review**

---

### 🚨 **High: Draining pods can become permanently unroutable if scale-down is canceled before the timeout**

In `pkg/controller/roleset/rolesyncer.go:475`, scale-down only calls `DeletePods` while `diff > 0`. `DeletePods` marks the selected pod with `aibrix.ai/draining=true` in `pkg/controller/drain/drain.go:105`, and gateway filtering then excludes it. If the user scales from 2 to 1, the pod starts draining; if they scale back to 2 before timeout, `diff == 0`, so no cleanup path runs and the pod keeps the drain annotations indefinitely. It remains Kubernetes-ready but is no longer routable. The same class of issue exists for direct PodSet scale-down at `pkg/controller/podset/podset_controller.go:337`.

**Expectation:** The controller should either clear drain annotations for pods no longer selected for deletion, or treat a draining pod as committed to deletion and create replacement capacity when desired replicas rise again.

**Affected locations:**

* `rolesyncer.go` (Line 475) — stateless scale-down logic.
* `PodSetReconciler.handleScaleDown` (`pkg/controller/podset/podset_controller.go:337`) — direct PodSet scale-down.

**Fix:** If a live pod still has drain annotations but is not in this reconcile’s delete set, strip the annotations and emit an event. Add a unit test for *"scale 2→1 starts drain, scale back to 2 clears annotations and does not delete."*

**Verification context:**

* Ran against a separate PR worktree at `/tmp/aibrix-pr2651.iQTxMj`.
* `go test ./pkg/controller/drain ./pkg/controller/podset ./pkg/utils` passed.
* `go test ./pkg/controller/roleset -run 'TestStatelessRoleSyncerScaleDown|TestStatelessRoleSyncer_Rollout|TestStatefulRoleSyncer|TestPodSetRoleSyncer'` passed.
* Broad envtest suites could not run locally because the required `bin/k8s/.../etcd` binaries are missing.

---

### 🚨 **Blocker: Drain is never cancelled (Original)**

`DeletePods` only starts or continues a drain. Nothing clears `aibrix.ai/draining=true` if that pod is no longer selected for deletion.

**Scenario (Scale 4→3, then back to 4 before the timeout):**

1. Scale-down marks pod A draining.
2. Scale-up sees `diff == 0` and does not put A into `toDelete`.
3. A keeps the annotation. The gateway keeps excluding it. Serving capacity stays at 3 until someone manually deletes the pod.

---

### ⚠️ **Should Fix: Not-ready pods still drain during rollout**

`processPod` skips the drain window only for `Succeeded` or `Failed` phases (`drain.go` Lines 97–107):

```go
if pod.Status.Phase == corev1.PodSucceeded || pod.Status.Phase == corev1.PodFailed {
	return deletePodsImmediately(...)
}
if annotations[aibrixconst.PodDrainingAnnotationKey] != "true" {
	return startDrain(...)
}

```

Rollout still sends not-ready / pending pods through `DeletePods` with `role.Drain` (for example, `StatefulRoleSyncer.Rollout` when `!IsPodReady`). Because those pods are not serving traffic anyway, enforcing the timeout only delays replacement—the exact same class of bug as the `ReplaceUnhealthy` issue, which correctly passes `nil`.

**Fix:** Pass `nil` drain at those call sites, or skip drain in `processPod` for pods that cannot take traffic.

---

### 👍 **What Looks Good**

* **API Design:** Optional API where `timeoutSeconds == 0` preserves today's delete-immediately behavior.
* **Gateway Integration:** Correctly wired: annotation patch $\rightarrow$ cache `updatePod` $\rightarrow$ `FilterReadyPod` / `FilterRoutablePods` $\rightarrow$ `selectTargetPod`.
* **Stateless Scale-down:** Prefers already-draining pods so it doesn't start a second concurrent drain.
* **Rolling Managers:** Properly wait on `Changed || RequeueAfter > 0`.
* **Propagation & Finalizers:** Drain config is copied onto generated `PodSets`, and `PodSet` finalizers drain member pods.
* **Resiliency:** Invalid drain states are reset instead of triggering immediate deletion, and events (`PodDrainStarted` / `PodDrainCompleted` / `PodDrainStateInvalid`) match the design doc.

---

### 📝 **Non-Blocking Notes**

* **PodSet vs. Direct-Pod Wait:** `PodSetRoleSyncer` deletes `PodSet` objects immediately and never returns `RequeueAfter`. Sequential wait then depends on `AllReady` seeing terminating `PodSets` plus pod watches. Drain duration for `podGroupSize > 1` is not as visible on the `RoleSet` result as direct-pod drain.
* **RoleSet Teardown:** `RoleSet.finalize` still deletes member pods immediately, whereas `PodSet` finalization does drain. Worth a one-line design doc note if `RoleSet` deletion is intentionally out of scope for v1.
* **Tests:** Envtest and e2e cover scale-down well; consider adding cancel-on-scale-up and a recreate-rollout drain case. Also note that webhook tests validate CRD OpenAPI (`minimum: 0`) via envtest rather than a validating webhook—which is totally fine, just slightly overstated in the PR body.
* **Nits:** `validateDrainState` still takes an unused `now` parameter, and the `FilterRoutablePods` godoc still omits mentioning draining pods.

**Review by varungup90 — APPROVED** [review]
